keps/sig-autoscaling/2021-scale-from-zero/README.md

```diff
@@ -0,0 +1,681 @@
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Currently disabling HPA is possible by manually setting the scaled resource to `replicas: 0`. This works as the HPA itself could never reach this state itself.
+As `replicas: 0` is now a possible state when using `minReplicas: 0` it can no longer be used to differentiate between manually disabled or automatically scaled to zero.
+
+Additionally the `replicas: 0` state is problematic as updating a HPA object `minReplicas` from `0` to `1` has different behavior. If `replicas` was `0` during the update, HPA
+will be disabled for the resource, if it was `> 0`, HPA will continue with the new `minReplicas` value.
+
+To resolve this issue the KEP is introducing an explicit `enableScaleToZero` property to explicitly enable/disable scale from/to zero.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+From an UX perspective the two stage opt-out / opt-in from scale to zero might feel a bit tedious, but the only other available option seems to be deprecating the implicit HPA pause on `replicas: 0`. While this might provide an improved
+UX, it would require a full deprecation cycle (12 months) before graduating this feature from alpha to beta.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We would add `EnableScaleToZero *bool` to the HPA `spec.behavior`.
+
+```golang
+type HorizontalPodAutoscalerBehavior struct {
+    ScaleUp           *HPAScalingRules
+    ScaleDown         *HPAScalingRules
+    EnableScaleToZero *bool
+}
+
+type HorizontalPodAutoscalerSpec struct {
+    ScaleTargetRef CrossVersionObjectReference
+    MinReplicas    *int32
+    MaxReplicas    int32
+    Metrics        []MetricSpec
+    Behavior       *HorizontalPodAutoscalerBehavior
+}
+```
+
+The `EnableScaleToZero` controls whether the `MinReplicas` can set to `>=0` (`true`, new behavior) or `>=1` (`false`, current behavior). The default will be `false` to preserve the current behavior.
+
+If `EnableScaleToZero` has been enabled, it can only be disabled when the scaled resource has at least one replica
+running and `MinReplicas` is `>=1`.
```

Comment:
Unless we differentiate between `EnableScaleToZero=false` and `EnableScaleToZero=nil`, the controller can't determine in this case whether this is the backwards compatible "pause" case or a valid transition, because if the HPA wouldn't be updated, `minReplicas: 1`, `replicas: 0` means HPA is stopped.

Reply:
I see the dilemma.  We can unset enable scale to zero and set min replicas = 1, but if the scale target ref is still at zero, it's not going to scale up again.

So we need a tri-state: /1/ old behavior (replicas=0 means disabled) /2/ scale to zero enabled and /3/ scale to zero disabled (but replicas=0 will immediately scale to 1).  Having different behavior for nil versus false sounds confusing.

There is a new KEP which adds `Spec.UpdatePolicy.UpdateMode` of which `Off` and `Auto` are dry-run and normal modes respectively: https://github.com/kubernetes/kubernetes/pull/107998.  Dry run mode is essentially the old behavior when replicas = 0.  We could make the default `UpdateMode` something like "ZeroMeansOff" instead of `Auto` so we can express all three states.

```golang
const (
	// UpdateModeZeroMeanOff (default) means the autoscaler never creates/deletes pods
	// when the scale target ref returns replicas = 0.
	UpdateModeZeroMeansOff = "ZeroMeansOff"
	// UpdateModeOff means that autoscaler never creates/deletes pods.
	// The HorizontalPodAutoscalerStatus.DesiredReplicas field is still populated.
	// This can be used for a "dry run"
	UpdateModeOff UpdateMode = "Off"
	// UpdateModeAuto means that autoscaler creates/deletes pods based on the scaling decision
	// including scaling to zero replicas.
	UpdateModeAuto UpdateMode = "Auto"
)
```

@liggitt @mwielgus what do you think about this API change, combining scale-to-zero and dry-run?

Reply:
Commented above, but removing the implicit pause on `replicas: 0` only for HPAs with `minReplicas: 0` makes more sense to me than a tri-state here.

Reply:
Okay. Let's keep the behavior as described then. You can resolve this comment.